src/nbl/ui/CWindowWin32.cpp

```diff
@@ -279,10 +310,65 @@ namespace ui
 				break;
 			case WA_INACTIVE:
 				eventCallback->onLostMouseFocus(window);
+				break;
+			}
+			break;
+		}
+		case WM_INPUT_DEVICE_CHANGE:
+		{
+			constexpr uint32_t CIRCULAR_BUFFER_CAPACITY = 256;
+			RID_DEVICE_INFO deviceInfo;
+			deviceInfo.cbSize = sizeof(RID_DEVICE_INFO);
+			UINT size = sizeof(RID_DEVICE_INFO);
+			GetRawInputDeviceInfoA((HANDLE)lParam, RIDI_DEVICEINFO, &deviceInfo, &size);
+
+			UINT deviceNameSize;
+			std::string deviceName;
+			if (GetRawInputDeviceInfo((HANDLE)lParam, RIDI_DEVICENAME, nullptr, &deviceNameSize) >= 0) {
+				deviceName.resize(deviceNameSize);
+				GetRawInputDeviceInfo((HANDLE)lParam, RIDI_DEVICENAME, (LPVOID)deviceName.data(), &deviceNameSize);
+			}
+
+			switch (wParam)
+			{
+			case GIDC_ARRIVAL:
+			{
+				if (deviceInfo.dwType == RIM_TYPEMOUSE)
+				{
+					//TODO: will finish that stuff tomorrow
+					auto chanel = core::make_smart_refctd_ptr<IMouseEventChannel>(CIRCULAR_BUFFER_CAPACITY);
+					decltype(s_mouseEventChannels)::value_type uniqueInputChannel;
+					uniqueInputChannel
+				}
+				else if (deviceInfo.dwType == RIM_TYPEKEYBOARD)
+				{
+
+				}
+				else if (deviceInfo.dwType == RIM_TYPEHID)
```

Comment:
whats TYPEHID?

Reply:
Any input device other than mouse/kb

Reply:
i guess we'll ignore this? what are planning to do with this? is there a way to detect game controllers getting connected (dont want it implemented yet, being curious)

Reply:
> i guess we'll ignore this? what are planning to do with this?

dunno, thought it's gonna be useful someday

> is there a way to detect game controllers getting connected (dont want it implemented yet, being curious)

Yeah, although processing game/other controllers is much more difficult than mouse/keyboard

Reply:
> dunno, thought it's gonna be useful someday

fair enough